lib/gibbon/export.rb

```diff
@@ -14,28 +14,47 @@ def export_api_url
       "http://#{get_data_center_from_api_key}api.mailchimp.com/export/1.0/"
     end
 
-    def call(method, params = {})
-      ensure_api_key params
+    # fsluis: Alternative, streaming, interface to mailchimp export api
+    #         Prevents having to keep lots of data in memory
+    def call(method, params = {}, &block)
+      rows = []
 
       api_url = export_api_url + method + "/"
       params = @default_params.merge(params).merge({:apikey => @api_key})
-      response = self.class.post(api_url, :body => MultiJson.dump(params), :timeout => @timeout)
-
-      lines = response.body.lines
-      if @throws_exceptions
-        # ignore blank responses
-        return [] if !lines.first || lines.first.strip.empty?
-
-        first_line = MultiJson.load(lines.first) if lines.first
+      block = Proc.new { |row| rows << row } unless block_given?
+      ensure_api_key params
 
-        if should_raise_for_response?(first_line)
-          error = MailChimpError.new(first_line["error"])
-          error.code = first_line["code"]
-          raise error
+      url = URI.parse(api_url)
+      req = Net::HTTP::Post.new(url.path, initheader = {'Content-Type' => 'application/json'})
+      req.body = MultiJson.dump(params)
+      puts params
+      Net::HTTP.start(url.host, url.port, :read_timeout => @timeout) do |http|
+        # http://stackoverflow.com/questions/29598196/ruby-net-http-read-body-nethttpokread-body-called-twice-ioerror
+        http.request req do |response|
+          i = -1
+          last = ''
+          response.read_body do |chunk|
+            #puts "Chunk length: #{chunk.length}"
+            #puts "Chunk: #{chunk}"
+            next if chunk.nil?
+            lines = (last+chunk).split("\n")
+            # There seems to be a bug (?) in the export API. Sometimes there's not a newline in between json objects...
+            lines = split_json(lines, '][')
```

Comment:
In theory this can be risky. When someone has the combinations of '][' or '}{' in their email addresses or click-urls (whether that's possible at all I don't know), this might cause a erroneous split. The risky part can be solved by checking whether the back-to-back bracket combinations are within a String or not.

But, it is not pretty for sure! This is a bug @ mailchimp export api, which should be dealt with by them accordingly - not here. For the time being however, this does form a possible workaround.

I see three options:
a) Leave it like this;
b) Add a check that the erroneous json code is not within a string;
c) Remove it.